status-im/status-legacy
# 5124 detect right to left text direction

[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

If you submit PR for issue with bounty then write here Fixes #NN where NN is issue number

*otherwise*

fixes #5124 

### Summary

[comment]: Problem : RTL languages are not properly identified. Needs enhancements.
...
Solution : 
1. The Regex constant containing the UTF-8 chars that are of RTL languages has been updated to include Kurdish, Urdu, Armenian, Farsi. Dhivehi is not included in UTF-8.
2. The android Native module 'text-direction' has been added and it invokes a builtin OS function that identifies which languages are present in a given string thus enabling to determine if a string is RTL or not. The language detection api exists on Android 8 only and does not work on Android 8.1 and 9. The language detection works fairly well, Officially it requires 5 words to work but it seems to get the correct results with 3 words in most cases.


### Review notes
All native modules return promises in react native and I'm trying to get the ClojureScript bit right but I get a message from Node during compilation that "UnhandledPromiseRejectionWarning: Unhandled promise rejection..." . I'm still checking this and it would be great of one of the Clojurescript guys at Status could help.
My code looks like this and to the bast of my understanding - nothing is missing (from src/status_im/chat/models/message_content.cljs) :
```

(when platform/android?
     (try
       (->
         ((.-isAvailable RNTextDirection)) ;1st neative module function
         (.then
           (fn [value]
             (when value
               (->
                 ((.-isRTL RNTextDirection) text)  ;2nd neative module function
                 (.then (fn [value] ( (or value (right-to-left-text? text) (assoc :rtl? true))))
                 (.catch (fn [error] ( (right-to-left-text? text) (assoc :rtl? true))))))))
         (.catch (fn [error] ( (right-to-left-text? text) (assoc :rtl? true)))))
       (catch js/Object err ( (right-to-left-text? text) (assoc :rtl? true))))))


```

#### Platforms
- Android
- iOS just to make sure nothing is broken.

##### Functional

- 1-1 chats
- public chats
- group chats

### Steps to test
Install the app, add a contact, start a chat with him/her and enter strings in Arabic/Hebrew/Farsi/Kurdish/Urdu/Dhivehi. Verify that they are displayed in the chat windows aligned as RTL.
